api/src/main/java/jakarta/json/bind/annotation/JsonbPolymorphicType.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.json.bind.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Configuration annotation of the polymorphic type handling.
+ * <br>
+ * This annotation is required to be present on the most top of all the classes,
+ * polymorphism should be applied to.
+ * <pre>{@code
+ * // Example
+ * @JsonbPolymorphicType(keyName = "@key")
+ * interface Animal {}
+ *
+ * class Dog implements Animal {}
+ * class Cat implements Animal {}
+ * }</pre>
+ * This annotation is closely tight with {@link JsonbSubtype}. It is recommended to use
+ * {@link JsonbSubtype} annotations to specify all the possible classes and their aliases.
+ * <br>
+ */
+@JsonbAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE})
+public @interface JsonbPolymorphicType {
+
+    /**
+     * Key name used for keeping polymorphic information when {@link Format#PROPERTY} is chosen.
+     *
+     * @return key name
+     */
+    String keyName() default "";
+
+    /**
+     * Specification of how the polymorphic information will be stored in the resulting JSON.
+     *
+     * @return polymorphic information storage type
+     */
+    Format format() default Format.PROPERTY;
```

Comment:
depending the outcome of the next comments we can drop it if we only support one format in next release